src/RDM/FrozenCoreRDMBuilder.cpp

```diff
@@ -0,0 +1,167 @@
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+// 
+// GQCG-gqcp is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with GQCG-gqcp.  If not, see <http://www.gnu.org/licenses/>.
+// 
+#include "RDM/FrozenCoreRDMBuilder.hpp"
+#include "utilities/linalg.hpp"
+
+
+namespace GQCP {
+
+/*
+ *  CONSTRUCTORS
+ */
+FrozenCoreRDMBuilder::FrozenCoreRDMBuilder(std::shared_ptr<BaseRDMBuilder> rdm_builder, size_t X) :
+        BaseRDMBuilder(),
+        rdm_builder (std::move(rdm_builder)),
+        X (X)
+{}
+
+
+
+/*
+ * OVERRIDDEN PUBLIC METHODS
+ */
+
+/**
+ *  @param x        the coefficient vector representing the wave function
+ *
+ *  @return all 1-RDMs given a coefficient vector
+ */
+OneRDMs FrozenCoreRDMBuilder::calculate1RDMs(const Eigen::VectorXd& x) const {
+
+    auto K = this->get_fock_space()->get_K();
+
+    Eigen::MatrixXd D_aa = Eigen::MatrixXd::Zero(K, K);
+    Eigen::MatrixXd D_bb = Eigen::MatrixXd::Zero(K, K);
+
+    auto Kn = K - this->X;
+
+    // frozen orbital values
+    for (size_t i = 0; i < this->X; i++) {
+        D_aa(i,i) = 1;
+        D_bb(i,i) = 1;
+    }
+
+    // Retrieve 1RDMs from non-frozen sub-space
+    OneRDMs sub_1rdms = this->rdm_builder->calculate1RDMs(x);
+
+    // Incorporate sub rdms into total rdms
+    D_aa.block(this->X, this->X, Kn, Kn) += sub_1rdms.one_rdm_aa.get_matrix_representation();
+    D_bb.block(this->X, this->X, Kn, Kn) += sub_1rdms.one_rdm_bb.get_matrix_representation();
+
+    OneRDM one_rdm_aa (D_aa);
+    OneRDM one_rdm_bb (D_bb);
+
+    return OneRDMs (one_rdm_aa, one_rdm_bb);
+};
+
+/**
+ *  @param x        the coefficient vector representing the wave function
+ *
+ *  @return all 2-RDMs given a coefficient vector
+ */
+TwoRDMs FrozenCoreRDMBuilder::calculate2RDMs(const Eigen::VectorXd& x) const {
+
+    auto K = this->get_fock_space()->get_K();
+
+    Eigen::Tensor<double, 4> d_aaaa (K,K,K,K);
+    d_aaaa.setZero();
+    Eigen::Tensor<double, 4> d_aabb (K,K,K,K);
+    d_aabb.setZero();
+    Eigen::Tensor<double, 4> d_bbaa (K,K,K,K);
+    d_bbaa.setZero();
+    Eigen::Tensor<double, 4> d_bbbb (K,K,K,K);
+    d_bbbb.setZero();
+
+    // Retrieve sub one RDMs to implement the frozen RDM formulas
+    OneRDMs one_rdms = this->rdm_builder->calculate1RDMs(x);
+    auto D_aa = one_rdms.one_rdm_aa.get_matrix_representation();
+    auto D_bb = one_rdms.one_rdm_bb.get_matrix_representation();
+
+    // Implement frozen RDM formulas
+    for (size_t p = 0; p < this->X; p++) {
+
+        GQCP::tensorBlockAddition<0,1>(d_aaaa, D_aa, this->X, this->X, p, p);
```

Comment:
Add one example which parts of the tensor you're setting, this is just a wall of code I don't understand immediately